index/load_doc.ml

```diff
@@ -1,47 +1,55 @@
+module Elt = Db.Elt
+module Db_common = Db
 module Types = Db.Types
-open Odoc_model
 module ModuleName = Odoc_model.Names.ModuleName
 
-let copy str = String.init (String.length str) (String.get str)
+let generic_cost ~ignore_no_doc name has_doc =
+  (* name length is important not because short identifier are better in the
+     abstract, but because the shortest result will be close to the query, as
+     the suffix tree does not return results shorter than the query*)
+  (String.length name * 6)
+  (* + (5 * List.length path) TODO : restore depth based ordering *)
+  + (if ignore_no_doc || has_doc then 0 else 30)
+  + if String.starts_with ~prefix:"Stdlib." name then -100 else 0
 
-let deep_copy (type t) (x : t) : t =
-  let buf = Marshal.(to_bytes x [ No_sharing; Closures ]) in
-  Marshal.from_bytes buf 0
+let type_cost paths =
+  paths |> List.concat |> List.map String.length |> List.fold_left ( + ) 0
 
-module Cache_doc = Cache.Make (struct
-  type t = Html_types.li_content_fun Tyxml.Html.elt
+let kind_cost (kind : Elt.Kind.t) =
+  match kind with
+  | Constructor type_path | Field type_path | Val type_path ->
+      type_cost type_path
+  | Doc -> 400
+  | TypeDecl _ | Module -> 0
+  | Exception _ | Class_type | Method | Class -> 10
+  | TypeExtension -> 1000
+  | ExtensionConstructor _ | ModuleType -> 10
 
-  let copy x = deep_copy x
-end)
-
-module Cache_name = Cache.Make (struct
-  type t = string
-
-  let copy = copy
-end)
-
-module Cache = Cache.Make (struct
-  type t = string
-
-  let copy = copy
-end)
+let cost ~name ~kind ~doc_html =
+  let ignore_no_doc =
+    match kind with
+    | Elt.Kind.Module | ModuleType -> true
+    | _ -> false
+  in
+  let has_doc = doc_html <> "" in
+  (* TODO : use entry cost *)
+  generic_cost ~ignore_no_doc name has_doc + kind_cost kind
 
-let clear () =
-  Cache.clear () ;
-  Cache_name.clear () ;
-  Cache_doc.clear ()
+(*
 
-let rec type_size = function
-  | Odoc_model.Lang.TypeExpr.Var _ -> 1
-  | Any -> 1
-  | Arrow (lbl, a, b) ->
-      (match lbl with
-      | None -> 0
-      | Some _ -> 1)
-      + type_size a + type_size b
-  | Constr (_, args) -> List.fold_left (fun acc t -> acc + type_size t) 1 args
-  | Tuple args -> List.fold_left (fun acc t -> acc + type_size t) 1 args
-  | _ -> 100
+  todo : check usefulness
```

Comment:
type_size is likely useful yes, I do not really know why it is commented anymore

Reply:
However there are concerns because not every entry has a type. It used to be the case but know that its not we need to think about how to handle untyped entries.

Reply:
We can also maybe ignore this because it works fine as far as I know

Reply:
Yes the type size was used as a malus and I felt it improved the search results (just like a `Very.Long.And_nested.Name.is_less_likely` to be what the user is looking for). Large function types have more potential to match the user requests, but we should prefer a shorter type if it exists! (the shorter type will be closer to what the user had in mind)

Reply:
Yes, but how do you compare an entry with a type with one that does not have a type ?

Reply:
I would say the type size is 0 in that case? (but that human testing is required to ensure it leads to a nice ordering of the results...)